Enable cd workflow

justusbunsi opened the pull request:
To reduce manual overhead on releases, this can be automated.

justusbunsi commented:
Referring to these docs: https://www.jenkins.io/doc/developer/publishing/releasing-cd/

lafriks commented:
I was wondering if there is really no way to release version only on tag and with tag number 🤔

lafriks commented:
I think you are missing `.github/workflows/cd.yaml` - template: https://github.com/jenkinsci/.github/blob/master/workflow-templates/cd.yaml

Maybe this template could be tuned to publish only tags and with tag as plugin version?

justusbunsi commented:
> I think you are missing `.github/workflows/cd.yaml` - template: https://github.com/jenkinsci/.github/blob/master/workflow-templates/cd.yaml
> 
> Maybe this template could be tuned to publish only tags and with tag as plugin version?

😆 Had this locally and didn't noticed that it somehow got ignored from being added.

justusbunsi commented:
Wouldn't the current workflow template create a tag and release on behalf due to "Release drafter"?

lafriks commented:
Oh, you are right, it will create tag/release from workflow 😞 

I would prefer not to release on every merged PR but manually

lafriks commented:
We still need to define secrets for Maven/Artifactory before merging

justusbunsi commented:
> We still need to define secrets for Maven/Artifactory before merging

This should be added automatically since the infra PR got merged.

lafriks commented:
yes, you are right

lafriks commented on .github/workflows/cd.yaml:
```diff
@@ -0,0 +1,57 @@
+name: cd
+on:
+  workflow_dispatch:
+  check_run:
+    types:
+      - completed
```
Let just use manual releases imho
```suggestion
```

justusbunsi commented:
> Oh, you are right, it will create tag/release from workflow 😞 
> 
> I would prefer not to release on every merged PR but manually

I'm not 100% sure if this really is the case. Regarding the docs there are some labels indicating merges "of interest". As long as there is none of these, the main branch doesn't get released. Maybe I am missing something.
Let's investigate the behaviour and adjust this PR according to our needs before merging. Maybe we can configure more than we see right now.

lafriks commented:
I still would prefer to use this part ;)
![attels](https://user-images.githubusercontent.com/165205/139707422-18e1290f-cf60-4439-ba7b-ebb29f40cd7a.png)

justusbunsi commented:
Do we keep the versioning as if it would be fully automated?

lafriks commented:
Yes, I don't think there is other option, imho. If there is maybe it can be somehow done using workflow inputs to specify version number: https://github.blog/changelog/2020-07-06-github-actions-manual-triggers-with-workflow_dispatch/

justusbunsi commented:
> If there is maybe it can be somehow done using workflow inputs to specify version number: https://github.blog/changelog/2020-07-06-github-actions-manual-triggers-with-workflow_dispatch/

This works pretty well. Have used it a few times. But not with Java projects.

lafriks commented:
that acutaly sounds good

justusbunsi commented:
You mean changing the versioning strategy or release with Gitea patches? Or both? 😀

lafriks commented:
I mean manually changing version and than manually triggering release workflow

lafriks commented:
We can still manually manage this part:
![attels](https://user-images.githubusercontent.com/165205/139849700-6802d3cd-8eae-4513-a402-ba6371649ec3.png)

justusbunsi commented:
@lafriks Apologies for the delayed response. 🙂 Concluding all the comments above it seems that we want to __manually create tag__ and then have the GitHub actions __automatically craft a release based on that tag__ and __automatically create a changelog__ using `release-drafter`.

We would have two options regarding version value in `pom.xml`. Not sure if any of both has disadvantage over the other one:

- Keep a placeholder and not update the version upon a release and just use the tag name during release build as version value
- Manually bump the version before tagging

---

Does this sums it up? If so, I'll modify the scripts according to that.

justusbunsi commented:
@lafriks Hrm... I struggle to understand how the described behavior can be integrated into the current workflow. There won't be automatic releases since the workflow only reacts to dispatch events, so we have full control over it right now. How about first see the current release in action before changing the automation?

As discussed above, I've changed the `pom.xml` to use the second approach from [the docs](https://www.jenkins.io/doc/developer/publishing/releasing-cd/#pom-file-modifications) but also dropped the maven.config change regarding changelist format. So it _should_ result in a proper 1.4.0. Maybe the tag name will differ from the others.

If this is ok for you, then let's find out what happens. 🙈

lafriks commented:
Ok, let's see what happens :)
